File: packages/backend/src/message.ts
import { Data } from 'ws';
import { Maybe, Types } from '@dotstats/common';

export function parseMessage(data: Data): Maybe<Message> {
  try {
    const message = JSON.parse(data.toString());

    if (message && typeof message.msg === 'string' && typeof message.ts === 'string') {
      message.ts = new Date(message.ts);

      return message;
    }
  } catch (_) {
    console.warn('Error parsing message JSON');
  }

  return null;
}

export function getBestBlock(message: Message): Maybe<BestBlock> {
  switch (message.msg) {
    case 'node.start':
    case 'system.interval':
    case 'block.import':
      return message;
    default:
      return null;
  }
}

interface MessageBase {
  ts: Date,
  level: 'INFO' | 'WARN',
}

export interface BestBlock {
  best: Types.BlockHash;
  height: Types.BlockNumber;
  ts: Date;
}

export interface AfgFinalized {
  ts: Date;
  finalized_number: Types.BlockNumber;
  finalized_hash: Types.BlockHash;
  msg: 'afg.finalized';
}

export interface AfgReceived {
  ts: Date;
  target_number: Maybe<Types.BlockNumber>;
  target_hash: Maybe<Types.BlockHash>;
  voter: Types.Address;
}

export interface AfgReceivedPrecommit extends AfgReceived {
  msg: 'afg.received_precommit';
}

export interface AfgReceivedPrevote extends AfgReceived {
  msg: 'afg.received_prevote';
}

export interface AfgReceivedCommit extends AfgReceived {
  msg: 'afg.received_commit';
}

export interface AfgAuthoritySet {
  msg: 'afg.authority_set';
  ts: Date;
  authority_id: Types.Address,
  authorities: Types.Authorities;
  authority_set_id: Types.AuthoritySetId;
  number: Types.BlockNumber;
  hash: Types.BlockHash;
}

export interface SystemConnected {
  msg: 'system.connected';
  name: Types.NodeName;
  chain: Types.ChainLabel;
  config: string;
  implementation: Types.NodeImplementation;
  version: Types.NodeVersion;
  authority: Maybe<boolean>;
  network_id: Maybe<Types.NetworkId>;
}

export interface SystemInterval extends BestBlock {
  msg: 'system.interval';
  network_state: Maybe<Types.NetworkState>;
  txcount: Types.TransactionCount;
  peers: Types.PeerCount;
  memory: Maybe<Types.MemoryUse>;
  cpu: Maybe<Types.CPUUse>;
  status: 'Idle' | string; // TODO: 'Idle' | ...?
  bandwidth_upload: Maybe<Types.BytesPerSecond>;
  bandwidth_download: Maybe<Types.BytesPerSecond>;
  finalized_height: Maybe<Types.BlockNumber>;
  finalized_hash: Maybe<Types.BlockHash>;
}

export interface SystemNetworkState extends MessageBase {
  msg: 'system.network_state';
  state: Types.NetworkState;
}

export interface NodeStart extends BestBlock {
  msg: 'node.start';
}

export interface BlockImport extends BestBlock {
  msg: 'block.import';
}

// Union type
export type Message = MessageBase & (
  | SystemConnected
  | SystemInterval
  | SystemNetworkState
  | NodeStart
  | BlockImport
  | AfgFinalized
  | AfgReceivedPrecommit
  | AfgReceivedPrevote
  | AfgReceivedCommit
  | AfgAuthoritySet
);


// received: {"msg":"block.import","level":"INFO","ts":"2018-06-18T17:30:35.285406538+02:00","best":"3d4fdc7960078ddc9be87dddc48324a6d64afdf1f65fffe89529ce9965cd5f29","height":526}
// received: {"msg":"node.start","level":"INFO","ts":"2018-06-18T17:30:40.038731057+02:00","best":"3d4fdc7960078ddc9be87dddc48324a6d64afdf1f65fffe89529ce9965cd5f29","height":526}
// received: {"msg":"system.connected","level":"INFO","ts":"2018-06-18T17:30:40.038975471+02:00","chain":"dev","config":"","version":"0.2.0","implementation":"parity-polkadot","name":"Majestic Widget"}
// received: {"msg":"system.interval","level":"INFO","ts":"2018-06-19T14:00:05.091355364+02:00","txcount":0,"best":"360c9563857308703398f637932b7ffe884e5c7b09692600ff09a4d753c9d948","height":7559,"peers":0,"status":"Idle"}

File: backend/core/src/node/message.rs
use crate::node::NodeDetails;
use crate::types::{Block, BlockHash, BlockNumber, ConnId};
use crate::util::Hash;
use actix::prelude::*;
use serde::de::IgnoredAny;
use serde::Deserialize;

#[derive(Deserialize, Debug, Message)]
#[rtype(result = "()")]
#[serde(untagged)]
pub enum NodeMessage {
    V1 {
        #[serde(flatten)]
        payload: Payload,
    },
    V2 {
        id: ConnId,
        payload: Payload,
    },
}

impl NodeMessage {
    /// Returns the connection ID or 0 if there is no ID.
    pub fn id(&self) -> ConnId {
        match self {
            NodeMessage::V1 { .. } => 0,
            NodeMessage::V2 { id, .. } => *id,
        }
    }
}

impl From<NodeMessage> for Payload {
    fn from(msg: NodeMessage) -> Payload {
        match msg {
            NodeMessage::V1 { payload, .. } | NodeMessage::V2 { payload, .. } => payload,
        }
    }
}

#[derive(Deserialize, Debug)]
#[serde(tag = "msg")]
pub enum Payload {
    #[serde(rename = "system.connected")]
    SystemConnected(SystemConnected),
    #[serde(rename = "system.interval")]
    SystemInterval(SystemInterval),
    #[serde(rename = "block.import")]
    BlockImport(Block),
    #[serde(rename = "notify.finalized")]
    NotifyFinalized(Finalized),
    #[serde(rename = "txpool.import")]
    TxPoolImport(IgnoredAny),
    #[serde(rename = "afg.finalized")]
    AfgFinalized(AfgFinalized),
    #[serde(rename = "afg.received_precommit")]
    AfgReceivedPrecommit(AfgReceivedPrecommit),
    #[serde(rename = "afg.received_prevote")]
    AfgReceivedPrevote(AfgReceivedPrevote),
    #[serde(rename = "afg.received_commit")]
    AfgReceivedCommit(AfgReceivedCommit),
    #[serde(rename = "afg.authority_set")]
    AfgAuthoritySet(AfgAuthoritySet),
    #[serde(rename = "afg.finalized_blocks_up_to")]
    AfgFinalizedBlocksUpTo(IgnoredAny),
    #[serde(rename = "aura.pre_sealed_block")]
    AuraPreSealedBlock(IgnoredAny),
    #[serde(rename = "prepared_block_for_proposing")]
    PreparedBlockForProposing(IgnoredAny),
}

#[derive(Deserialize, Debug)]
pub struct SystemConnected {
    pub genesis_hash: Hash,
    #[serde(flatten)]
    pub node: NodeDetails,
}

#[derive(Deserialize, Debug)]
pub struct SystemInterval {
    pub peers: Option<u64>,
    pub txcount: Option<u64>,
    pub bandwidth_upload: Option<f64>,
    pub bandwidth_download: Option<f64>,
    pub finalized_height: Option<BlockNumber>,
    pub finalized_hash: Option<BlockHash>,
    #[serde(flatten)]
    pub block: Option<Block>,
    pub used_state_cache_size: Option<f32>,
}

#[derive(Deserialize, Debug)]
pub struct Finalized {
    #[serde(rename = "best")]
    pub hash: BlockHash,
    pub height: Box<str>,
}

#[derive(Deserialize, Debug)]
pub struct AfgAuthoritySet {
    pub authority_id: Box<str>,
    pub authorities: Box<str>,
    pub authority_set_id: Box<str>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct AfgFinalized {
    pub finalized_hash: BlockHash,
    pub finalized_number: Box<str>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct AfgReceived {
    pub target_hash: BlockHash,
    pub target_number: Box<str>,
    pub voter: Option<Box<str>>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct AfgReceivedPrecommit {
    #[serde(flatten)]
    pub received: AfgReceived,
}

#[derive(Deserialize, Debug, Clone)]
pub struct AfgReceivedPrevote {
    #[serde(flatten)]
    pub received: AfgReceived,
}

#[derive(Deserialize, Debug, Clone)]
pub struct AfgReceivedCommit {
    #[serde(flatten)]
    pub received: AfgReceived,
}

impl Block {
    pub fn zero() -> Self {
        Block {
            hash: BlockHash::from([0; 32]),
            height: 0,
        }
    }
}

impl Payload {
    pub fn best_block(&self) -> Option<&Block> {
        match self {
            Payload::BlockImport(block) => Some(block),
            Payload::SystemInterval(SystemInterval { block, .. }) => block.as_ref(),
            _ => None,
        }
    }

    pub fn finalized_block(&self) -> Option<Block> {
        match self {
            Payload::SystemInterval(ref interval) => Some(Block {
                hash: interval.finalized_hash?,
                height: interval.finalized_height?,
            }),
            Payload::NotifyFinalized(ref finalized) => Some(Block {
                hash: finalized.hash,
                height: finalized.height.parse().ok()?,
            }),
            _ => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn message_v1() {
        let json = r#"{"msg":"notify.finalized","level":"INFO","ts":"2021-01-13T12:38:25.410794650+01:00","best":"0x031c3521ca2f9c673812d692fc330b9a18e18a2781e3f9976992f861fd3ea0cb","height":"50"}"#;
        assert!(
            matches!(
                serde_json::from_str::<NodeMessage>(json).unwrap(),
                NodeMessage::V1 { .. },
            ),
            "message did not match variant V1",
        );
    }

    #[test]
    fn message_v2() {
        let json = r#"{"id":1,"ts":"2021-01-13T12:22:20.053527101+01:00","payload":{"best":"0xcc41708573f2acaded9dd75e07dac2d4163d136ca35b3061c558d7a35a09dd8d","height":"209","msg":"notify.finalized"}}"#;
        assert!(
            matches!(
                serde_json::from_str::<NodeMessage>(json).unwrap(),
                NodeMessage::V2 { .. },
            ),
            "message did not match variant V2",
        );
    }
}
